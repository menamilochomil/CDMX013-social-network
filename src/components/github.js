/* eslint-disable import/no-unresolved */
import {
  getAuth, GithubAuthProvider, signInWithRedirect, getRedirectResult,
} from '../lib/imports.js';
import { app } from '../lib/firebase.js';
import { onNavigate } from '../main.js';

const githubProvider = new GithubAuthProvider();
const auth = getAuth(app);

export const githubLogin = () => {
  signInWithRedirect(auth, githubProvider);
  getRedirectResult(auth)
    .then((result) => {
      const credential = GithubAuthProvider.credentialFromResult(result);
      if (credential) {
        // const token = result.credential.accessToken;
        onNavigate('/home');
      }
      // const user = result.user;
    })
    .catch((err) => {
      console.log(err);
    });
};
