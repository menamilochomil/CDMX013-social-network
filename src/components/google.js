/* eslint-disable no-unused-vars */
/* eslint-disable import/no-unresolved */
import { GoogleAuthProvider, signInWithPopup, getAuth } from '../lib/imports.js';
import { onNavigate } from '../main.js';

export const googleAuth = async () => {
  const auth = getAuth();
  const provider = new GoogleAuthProvider();
  await signInWithPopup(auth, provider)
    .then((result) => {
      const credential = GoogleAuthProvider.credentialFromResult(result);
      const token = credential.accessToken;
      onNavigate('/home');
      console.log('google sign in');
    })
    .catch((err) => {
      console.log(err);
    });
};
