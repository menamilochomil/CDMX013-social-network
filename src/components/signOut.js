/* eslint-disable import/no-unresolved */
import { getAuth, signOut } from 'https://www.gstatic.com/firebasejs/9.9.3/firebase-auth.js';
import { app } from '../lib/firebase.js';
import { onNavigate } from '../main.js';

const auth = getAuth(app);

export const signout = () => {
  signOut(auth).then(() => {
    onNavigate('/');
  });
};
