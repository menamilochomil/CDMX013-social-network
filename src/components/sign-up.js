/* eslint-disable no-unused-vars */
/* eslint-disable import/no-unresolved */
import { getAuth, createUserWithEmailAndPassword } from '../lib/imports.js';
import { app } from '../lib/firebase.js';
import { onNavigate } from '../main.js';
import { githubLogin } from './github.js';

export const auth = getAuth(app);

export const signUp = () => {
  const divContainer = document.createElement('div');
  const logo = document.createElement('img');
  const divInputs = document.createElement('div');
  const p = document.createElement('p');
  const boxEmail = document.createElement('input');
  const boxPassword = document.createElement('input');
  const boxConfirmPassword = document.createElement('input');
  const signUpButton = document.createElement('button');
  const pMessage = document.createElement('p');
  const sectionOr = document.createElement('img');
  const signUpTwitter = document.createElement('img');
  const signUpGitHub = document.createElement('img');
  const signUpGoogle = document.createElement('img');
  const footer = document.createElement('footer');

  const paraError = document.createElement('p');
  const paraCongrats = document.createElement('p');

  logo.src = '../imgs/logo.png';
  logo.classList.add('logoTech');
  divInputs.setAttribute('class', 'containerInputs');
  p.textContent = 'Sing up';
  p.setAttribute('id', 'text');
  boxEmail.setAttribute('type', 'email');
  boxEmail.placeholder = 'email@something.com';
  boxEmail.setAttribute('class', 'inputs');
  boxEmail.setAttribute('id', 'email');
  boxPassword.setAttribute('type', 'password');
  boxPassword.placeholder = 'password';
  boxPassword.setAttribute('class', 'inputs');
  boxPassword.setAttribute('id', 'password');
  boxConfirmPassword.setAttribute('type', 'password');
  boxConfirmPassword.placeholder = 'Confirm your password';
  boxConfirmPassword.setAttribute('class', 'inputs');
  pMessage.textContent = ' By clicking “Sign up”, you agree to our terms of service, privacy policy and cookie policy.';
  pMessage.setAttribute('id', 'pMessage');
  signUpButton.textContent = 'Sign Up';
  signUpButton.setAttribute('class', 'purpleButton');
  signUpButton.setAttribute('id', 'sign-up');

  divInputs.append(boxEmail, boxPassword, boxConfirmPassword, paraError, paraCongrats, signUpButton, pMessage);

  sectionOr.src = '../imgs/sectionOr.png';
  signUpTwitter.src = '../imgs/Twitter.png';
  signUpTwitter.setAttribute('class', 'signUpIcon');
  signUpGitHub.src = '../imgs/Github.png';
  signUpGitHub.setAttribute('class', 'signUpIcon');
  signUpGoogle.src = '../imgs/Google.png';
  signUpGoogle.setAttribute('class', 'signUpIcon');
  footer.textContent = '2022';
  paraError.setAttribute('class', 'errorMessage');
  paraCongrats.setAttribute('id', 'congrats');

  function congrats() {
    onNavigate('/check');
  }

  const createAccount = async () => {
    const signUpEmail = boxEmail.value;
    const signUpPassword = boxPassword.value;
    const confirmPasword = boxConfirmPassword.value;

    try {
      if (signUpPassword !== confirmPasword) throw Error('The password does not match');
      const userCredencial = await createUserWithEmailAndPassword(auth, signUpEmail, signUpPassword, confirmPasword);
      const user = userCredencial.user;
      paraError.innerHTML = '';
      setTimeout(congrats, 1000);
    } catch (error) {
      paraError.innerHTML = error;

      if (error.code === 'auth/invalid-email') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'Your email is incorrect';
      }
      if (error.code === 'auth/email-already-in-use') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'The e-mail already exist.';
      }

      if (error.code === 'auth/weak-password') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'Your password should be at least 6 characters';
      }
      if (error.code === 'auth/network-request-failed') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'Connection failed';
      }

      if (signUpEmail === '' && signUpPassword === '' && confirmPasword === '') {
        paraError.innerHTML = 'Please, fill all the fields';
      } else if (boxEmail.value === '') {
        paraError.innerHTML = 'Please write an e-mail';
      } else if (boxPassword.value === '') {
        paraError.innerHTML = 'Please write a password';
      } else if (boxConfirmPassword.value === '') {
        paraError.innerHTML = 'Please confirm your password';
      }
    }
  };

  signUpButton.addEventListener('click', createAccount);

  logo.addEventListener('click', () => {
    onNavigate('/');
  });

  signUpGitHub.addEventListener('click', githubLogin);

  divContainer.append(
    logo,
    p,
    divInputs,
    signUpGitHub,
    footer,
  );

  divContainer.setAttribute('class', 'container');

  return divContainer;
};
