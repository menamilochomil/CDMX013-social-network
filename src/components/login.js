/* eslint-disable no-unused-vars */
/* eslint-disable import/no-unresolved */
import { getAuth, signInWithEmailAndPassword } from 'https://www.gstatic.com/firebasejs/9.9.3/firebase-auth.js';
import { onNavigate } from '../main.js';
import { app } from '../lib/firebase.js';
import { githubLogin } from './github.js';
import { googleAuth } from './google.js';

export const auth = getAuth(app);

export const login = () => {
  const divContainer = document.createElement('div');
  const logo = document.createElement('img');
  const divInputs = document.createElement('div');
  const p = document.createElement('p');
  const boxEmail = document.createElement('input');
  const boxPassword = document.createElement('input');
  const loginButton = document.createElement('button');
  const pAccount = document.createElement('p');
  const signUpButton = document.createElement('button');
  const divAccount = document.createElement('div');
  const sectionOr = document.createElement('img');
  const loginTwitter = document.createElement('img');
  const loginGitHub = document.createElement('img');
  const loginGoogle = document.createElement('img');
  const footer = document.createElement('footer');
  const paraError = document.createElement('p');

  logo.src = '../imgs/logo.png';
  logo.classList.add('logoTech');
  divInputs.setAttribute('class', 'containerInputs');
  p.textContent = 'Log in';
  p.setAttribute('class', 'text');

  boxEmail.setAttribute('type', 'email');
  boxEmail.placeholder = 'email@something.com';
  boxEmail.setAttribute('class', 'inputs');
  boxPassword.setAttribute('type', 'password');
  boxPassword.placeholder = 'password';
  boxPassword.setAttribute('class', 'inputs');
  loginButton.textContent = 'Log in';
  loginButton.setAttribute('class', 'purpleButton');

  divInputs.append(boxEmail, boxPassword, paraError, loginButton);

  pAccount.textContent = ' Do not you have an account yet? Please,  ';
  pAccount.setAttribute('id', 'pAccount');
  signUpButton.textContent = 'Sign up here ';
  divAccount.append(pAccount, signUpButton);
  divAccount.setAttribute('id', 'divAccount');
  signUpButton.setAttribute('id', 'signUpButton');
  sectionOr.src = '../imgs/sectionOr.png';
  loginTwitter.src = '../imgs/loginTwitter.png';
  loginTwitter.setAttribute('class', 'loginIcon');
  loginGitHub.src = '../imgs/loginGithub.png';
  loginGitHub.setAttribute('class', 'loginIcon');
  loginGoogle.src = '../imgs/loginGoogle.png';
  loginGoogle.setAttribute('class', 'loginIcon');
  footer.textContent = '2022';
  paraError.setAttribute('class', 'errorMessage');

  const firebaseLogIn = async () => {
    const loginEmail = boxEmail.value;
    const loginPassword = boxPassword.value;

    try {
      const userCredencial = await signInWithEmailAndPassword(
        auth,
        loginEmail,
        loginPassword,
      );
      const user = userCredencial.user;
      onNavigate('/home');
    } catch (error) {
      if (error.code === 'auth/user-not-found') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'You do not have an account yet';
      }
      if (error.code === 'auth/internal-error') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'Please, write your password';
      }
      if (error.code === 'auth/invalid-email') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'Your email is incorrect';
      }
      if (error.code === 'auth/wrong-password') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'Your password is incorrect';
      }
      if (error.code === 'auth/user-disabled') {
        paraError.style.display = 'block';
        paraError.style.opacity = '1';
        paraError.innerHTML = 'Your account is disabled';
      }
    }
  };

  loginButton.addEventListener('click', firebaseLogIn);

  logo.addEventListener('click', () => {
    onNavigate('/');
  });

  loginGoogle.addEventListener('click', googleAuth);

  signUpButton.addEventListener('click', () => {
    onNavigate('/signUp');
  });

  loginGitHub.addEventListener('click', githubLogin);

  divContainer.append(
    logo,
    p,
    divInputs,
    divAccount,
    sectionOr,
    loginGoogle,
    loginGitHub,
    footer,
  );

  divContainer.setAttribute('class', 'container');

  return divContainer;
};
