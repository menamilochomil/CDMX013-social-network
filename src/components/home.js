/* eslint-disable no-alert */
import {
  auth,
  publish,
  onGetPost,
  deletePost,
  like,
  dislike,
  getSomething,
} from '../lib/firebase.js';
import { signout } from './signOut.js';

export const home = () => {
  const divContainer = document.createElement('div');
  const header = document.createElement('header');
  const logoHome = document.createElement('img');
  const logout = document.createElement('button');
  const sectionContainer = document.createElement('section');
  const postForm = document.createElement('form');
  const nameUser = document.createElement('h1');
  const commentContainer = document.createElement('div');
  const writeComment = document.createElement('textarea');
  const charCounter = document.createElement('p');
  const errorCharacters = document.createElement('p');
  const shareButton = document.createElement('button');
  const comment = document.createElement('section');

  divContainer.setAttribute('id', 'body-home');
  header.setAttribute('id', 'header');
  logoHome.src = '../imgs/logo2.png';
  logoHome.setAttribute('id', 'logo-home');
  logout.textContent = 'Log out';
  logout.setAttribute('id', 'logout');
  sectionContainer.setAttribute('class', 'section-home');
  postForm.setAttribute('class', 'section-home');
  nameUser.setAttribute('id', 'name-user');
  commentContainer.setAttribute('id', 'comment-container');
  writeComment.setAttribute('id', 'post-container');
  writeComment.setAttribute(
    'placeholder',
    'Share something less than 500 characters with the community ...',
  );
  writeComment.setAttribute('maxlength', 500);
  charCounter.setAttribute('id', 'counter');
  shareButton.textContent = 'Share';
  shareButton.setAttribute('id', 'share');
  shareButton.disabled = true;

  header.append(logoHome, logout);
  commentContainer.appendChild(writeComment);
  postForm.append(nameUser, commentContainer, errorCharacters, shareButton);
  sectionContainer.append(postForm, comment);

  logout.addEventListener('click', () => {
    signout();
  });

  const userAuth = auth.currentUser;

  onGetPost((querySnapshot) => {
    const user = auth.currentUser;
    nameUser.textContent = `Welcome ${user.email} !`;
    let html = '';
    let htmlDelete = '';
    let htmlLike = '';
    querySnapshot.forEach((doc) => {
      const postData = doc.data();
      if (user.email === postData.user) {
        htmlDelete = `
          <i class="fa-sharp fa-solid fa-trash" data-id="${doc.id}"></i>`;
      } else {
        htmlDelete = '';
      }

      if (doc.data().likes.includes(user.email)) {
        htmlLike = `
        <i class="fa-solid fa-heart" data-id="${doc.id}"></i>
        `;
      } else {
        htmlLike = `
        <i class="fa-regular fa-heart" data-id="${doc.id}"></i>
        `;
      }
      let numberOflikes = '';
      if (doc.data().likes.length > 0) {
        numberOflikes = `<span>${doc.data().likes.length}</span>`;
      }
      html += `
          <article class="post-content">
          <div class= "container-header-post">
            <img id= "profile-home" src="../imgs/profile.png" alt="Profile Image">
            <h2 id ="header-post">${postData.user}</h2>
            <span id="delete-edit"> ${htmlDelete} </span>
          </div>
          <p class="post-user">${postData.posts}</p>
          <div class = "like-comment">
          <span class = "icon-container">
            ${htmlLike}
            <p>${numberOflikes} Like</p>
          </span>
          </div>
          </article>
          `;
    });
    comment.innerHTML = html;

    // Delete post
    const deteleTrash = comment.querySelectorAll('.fa-trash');

    deteleTrash.forEach((trash) => {
      trash.addEventListener('click', ({ target: { dataset } }) => {
        if (window.confirm('Do you really want to delete?')) {
          deletePost(dataset.id);
        }
      });
    });

    // Like post
    const likesIcons = comment.querySelectorAll('.fa-heart');

    likesIcons.forEach((item) => {
      item.addEventListener('click', async ({ target: { dataset } }) => {
        console.log(dataset.id);
        const doc = await getSomething(dataset.id);
        if (doc.data().likes.includes(user.email)) {
          dislike(dataset.id, user.email);
        } else {
          like(dataset.id, user.email);
        }
      });
    });
  });

  writeComment.addEventListener('input', (event) => {
    const { value } = event.target;
    if (value.length > 0) {
      charCounter.textContent = `${value.length}/500`;
      commentContainer.appendChild(charCounter);
    } else {
      charCounter.textContent = '';
    }
    if (value !== '') {
      shareButton.disabled = false;
    } else {
      shareButton.disabled = true;
    }
  });

  shareButton.addEventListener('click', (e) => {
    e.preventDefault();
    const postUser = postForm['post-container'];
    const userID = userAuth.email;
    publish(postUser.value, userID);
    postForm.reset();
    charCounter.textContent = '';
    shareButton.disabled = true;
  });

  divContainer.append(header, sectionContainer);
  return divContainer;
};
