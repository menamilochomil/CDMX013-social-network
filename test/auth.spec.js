/**
 * @jest-environment jsdom
 */
import { signUp } from '../src/components/sign-up';
// import { onNavigate } from '../src/main';

jest.mock('../src/lib/imports.js');
jest.mock('../src/main.js');

describe('The authentication of sign up', () => {
  test('this is a function', () => {
    expect(typeof signUp).toBe('function');
  });

  test('There exists an input email, password, confirm password', () => {
    const element = signUp();
    const inputs = element.querySelector('.inputs');
    expect(inputs).not.toBeNull();
  });

  test('It exists a sing up button', () => {
    const element = signUp();
    const btn = element.querySelector('#sign-up');
    expect(btn).not.toBeNull();
  });

  // test('It exists a sign up button', () => {
  //   const element = signUp();
  //   const email = element.querySelector('#email');
  //   const password = element.querySelector('#password');
  //   const btn = element.querySelector('#sign-up');
  //   const redirect = onNavigate('/check');
  //   email.value = 'example@gmail.com';
  //   password.value = '123456';
  //   btn.click();

  //   expect(redirect).not.toBeNull();
  // });

  test('Snapshot of sign up', () => {
    expect(signUp()).toMatchSnapshot();
  });
});
