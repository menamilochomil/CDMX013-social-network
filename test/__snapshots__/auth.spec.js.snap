// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The authentication of sign up Snapshot of sign up 1`] = `
<div
  class="container"
>
  <img
    class="logoTech"
    src="../imgs/logo.png"
  />
  <p
    id="text"
  >
    Sing up
  </p>
  <div
    class="containerInputs"
  >
    <input
      class="inputs"
      id="email"
      placeholder="email@something.com"
      type="email"
    />
    <input
      class="inputs"
      id="password"
      placeholder="password"
      type="password"
    />
    <input
      class="inputs"
      placeholder="Confirm your password"
      type="password"
    />
    <p
      class="errorMessage"
    />
    <p
      id="congrats"
    />
    <button
      class="purpleButton"
      id="sign-up"
    >
      Sign Up
    </button>
    <p
      id="pMessage"
    >
       By clicking “Sign up”, you agree to our terms of service, privacy policy and cookie policy.
    </p>
  </div>
  <img
    class="signUpIcon"
    src="../imgs/Github.png"
  />
  <footer>
    2022
  </footer>
</div>
`;
